### ISSUE TITLE
Great!!!  Useful!!!   Trying ...

### ISSUE THREAD

**Issue by yuedajiong** [konstantinosKokos/ape#5]
在Transformer中，位置编码和注意力部分:  
    1. 早期原生vanilla的版本，主要用傅里叶位置编码sin/cos，或者learnable的位置编码，然后add到token embedding中(也有cat而不是add)，然后做q,k,v相同的self-attention;  
    2. 而后来比如2025时代，如qwen3等，都是将position一路带入到attention计算块，作QKV中qk的独立输入；
    3. 编码策略也有很多独立的研究，在sin/cos, learnable之外，比如RoPE，GridPE，更有AlgebraicPE来一统Seq, Grid, Tree（没有Graph)等。
    4.  在我个人的一些研究中，还存在一种反向的需求，一些天生的弱seq的数据（弱的意思是虽然没有显示的附加position信息，但数据的顺序存储和网络本身的sequential性质赋予了seq性质），还需要通过池化性质的网络和特殊的本质带有池化性质的损失函数，来退化到set来预测，效果更好。
    5. 在我个人的一些研究中，一个体会就是，对于输入数据，很可能在nn的forward过程中，需要将数据的每个原子单位（比如pixel, token/char)的三种成分/factor（form/semantic/position)一直保持，form/position特别需要保持在原始空间。

**Comment by yuedajiong** [comment]
Can we NOT recompute these lines every forward in AlgebraicCCT@cct.py, because the h and w is fixed?

        h, w = patch_values.shape[-2], patch_values.shape[-2]
        patch_values = patch_values.permute(0, 2, 3, 1).flatten(1, 2)  #(-1, H_out * W_out, dim)
        x_pos = torch.arange(0, h, device=patch_values.device).unsqueeze(-1).expand(h, w).flatten()[None]
        y_pos = torch.arange(0, w, device=patch_values.device).unsqueeze(+0).expand(h, w).flatten()[None]
        self.positional_encoder.precompute(max(h, w))
        x_maps, y_maps = self.positional_encoder.forward(x_pos, y_pos)
        coords = torch.cat((x_pos, y_pos), dim=0)
        dists = ((coords[:, None] - coords[:, :, None]) ** 2).sum(0).sqrt().floor()[None, :, :, None, None]

**Comment by konstantinosKokos** [comment]
That's correct. You only need to recompute them if you are training the positional encoder primitives. Even in this case, you can do something smarter than what I've done, for example you can cache the representations and only do an optimization step on the positional encoder every N optimization steps of the full network (optionally with gradient accumulation). 

In inference, you can basically just precompute the entire `atn_fn` and the use it for all batches / images (as you can see, to you don't need any image input to make `atn_fn`)

**Comment by yuedajiong** [comment]
### command line
PYTHONPATH=. python scripts/image.py --model Algebraic --data_dir ./data/ --store_path ./outs/ --seed 1312  #--dataset cifar10
   a.  PYTHONPATH=.
   b.  #--dataset cifar10 

### bug about Sequential in cct.py
   from torch.nn import Module, Conv2d, Linear, MaxPool2d, Sequential, ReLU, Embedding
   from ape.nn.position import Grid, SinusoidalGrid, Sequential
   self.patch_embed = Sequential(     #modify: torch.nn.Sequential(

**Comment by yuedajiong** [comment]
A tidied mimi version for study:

[ape.py.txt](https://github.com/user-attachments/files/21880636/ape.py.txt)
